### ISSUE TITLE
Confused about the PAL loss on the outlier pixels.

### ISSUE THREAD

**Issue by gaozhitong** [tianyu0207/PEBAL#18]
Hi, thanks for your excellent work! I am a little confused about the PAL loss, especially on how it works on the outlier pixels in D_out.

Main Confusion: 
In eq(3) of the paper, as I understand, the label y_w should be Y+1 for outlier pixels. Then the loss term inside the log becomes f(Y+1;x)_w  + f(Y+1;x)_w / a_w. However, when I checked the code, I found that the real implementation of the PAL loss on the outlier pixels (inside log) seems to be a sum of the interior class probabilities and f(Y+1;x)_w / a_w. As can be seen in `reserve_boosting_energy = torch.add(true_pred, reservation.unsqueeze(1))[mask.unsqueeze(1).
                repeat(1, 19, 1, 1)].log()`.  Would you kindly explain this? (p.s. I have also read the paper [33]. It seems that they do not consider an outlier dataset during the training, therefore, I can not find helpful information.)

Others:
I find the code conducts the 'log' operation twice. (Besides the above code, another 'log' operation is conducted on `reserve_boosting_energy = torch.clamp(reserve_boosting_energy, min=1e-7).log()`). It is really confusing.
Also, there seems to be a minor typo in eq(3) where a softmax operation is lacking on the logit.

Your response would be fully appreciated!

**Comment by yyliu01** [comment]
Hi @gaozhitong , thanks for your interest! I apologise for the late reply.

For your main confusion, $y_w$ in here denotes the pixel-wise anotated (panel) label, where $w$ is the pixel location. The meaning for the equation **is very similar as the eq(6)** in gambler loss [33] (page 4, section 4.3, as shown in below):
![image](https://user-images.githubusercontent.com/102338056/186808372-17804a7c-2ff9-4117-ab47-f680b57b38e4.png)

In our cases, for the outlier pixel (after the anomaly-mix from coco), you'll see something like this for the input:

![image](https://user-images.githubusercontent.com/102338056/186808756-0d6c58cb-c0f9-4a12-807c-9172d1e8c53b.png)

In order to let the model aware those anomalies (pizza) regions, we boosting the model to put more logits (or confidence, after softmax) to the extra channel via a  smaller reward (described in [33], page 13, the last theorem proof). To simplify, the implementation is  **-log(P_inlier-channels + P_extra-channel/reward) * y'**, where y'=1, which means we penalize all inlier channels (samilar as negative log likelihood). In the early stage, the model will tend to put more logits in head classes (e.g., road), therefor other channels will be penalized. In the ideal situation, the model will put all the logits in the extra channel so that it will be no penalties from the loss. 

For the others confusion, yes, we make a typo when we re-organise the code, and it is fixed now. We do agree it should be more clear to add a softmax operation within the equation, but it will be a bit complexed to read the formula. Thus I'll add one more sentence in the arxiv version to avoid the misunderstanding for the other followers. Thanks so much for reporting it.

Hope that can make it more clear.

Best regards,
Yuyuan

**Comment by gaozhitong** [comment]
Hi, thanks for your very detailed reply!

After reading the explanation and theorem proof in [33] on page 13, I understand the intuition of what is done on anomaly regions. However, I still can not figure out **how equation (3) results in the implementation mentioned**. As mentioned in the paper, fθ(yω; x)ω represents the logit (or probability) for class yω at pixel ω, and yω = Y+1 for anomaly regions, then fθ(yω; x)ω should be the logit  (or probability) on the extra channel at pixel ω, instead of the summation of the interior channels. 
(p.s. I do understand how the loss works in normal regions and the implementation also conforms to the equation.)

Thanks again for your explanation and quick response. Looking forward to your reply!

**Comment by yyliu01** [comment]
Hi,

Sorry, I'm still confused about your question. 

The $y_{in}$ in here should indicate the inlier (channels') logits, while the $y_{out}$ is the inlier (channels') logits+extra channel. They belong to the same pixel,  at pixel $w$. 

Again, the equation is highly borrowed from [33] (as shown in below):
![image](https://user-images.githubusercontent.com/102338056/186927280-2e6c25ab-200b-465a-80cd-257ad0a72f4b.png)

The implementation is also similar in this line:
https://github.com/Z-T-WANG/NIPS2019DeepGamblers/blob/0d1b595611a8bc653fddfdf1419bd8dbde153532/main.py#L237

Please give me more details if this can't answer your question.

Regards,

**Comment by gaozhitong** [comment]
Hi,
To be more specific, I think the implementation of calculating **gambler_loss_out** from the codes (see following) differs from the paper, as well as [33]. 
(I do agree with the implementation of calculating  **gambler_loss_in**,  which is exactly the same with [33])

![image](https://user-images.githubusercontent.com/40078173/187019664-972f7926-50d6-4610-9601-d3c7e6884c24.png)
![image](https://user-images.githubusercontent.com/40078173/187019707-5dae1c75-46cb-4525-81f2-095a113d5c0e.png)

**Comment by tianyu0207** [comment]
> Hi, To be more specific, I think the implementation of calculating **gambler_loss_out** from the codes (see following) differs from the paper, as well as [33]. (I do agree with the implementation of calculating **gambler_loss_in**, which is exactly the same with [33])
> 
> ![image](https://user-images.githubusercontent.com/40078173/187019664-972f7926-50d6-4610-9601-d3c7e6884c24.png) ![image](https://user-images.githubusercontent.com/40078173/187019707-5dae1c75-46cb-4525-81f2-095a113d5c0e.png)

Hi Sorry for the confusion. We have fixed the texts that may lead to misunderstandings in the Arxiv version. I'm closing the issues now and feel free to reopen it if needed.
